app/models/manageiq/providers/amazon/agent_coordinator.rb

```diff
@@ -144,68 +131,92 @@ def deploy_agent
     security_group_id = find_or_create_security_group(subnets[0].vpc_id)
     find_or_create_profile
 
-    instance                = ec2.create_instances(
+    instance = ec2.create_instances(
       :image_id             => get_agent_image_id,
       :min_count            => 1,
       :max_count            => 1,
       :key_name             => kp.name,
-      :security_group_ids   => [security_group_id],
       :instance_type        => 't2.micro',
       :placement            => {
         :availability_zone => zone_name
       },
-      :subnet_id            => subnets[0].subnet_id,
       :iam_instance_profile => {
         :name => label
       },
+      :network_interfaces   => [{
+        :associate_public_ip_address => true,
+        :delete_on_termination       => true,
+        :device_index                => 0,
+        :subnet_id                   => subnets[0].subnet_id,
+        :groups                      => [security_group_id]
+      }],
       :tag_specifications   => [{
         :resource_type => "instance",
         :tags          => [{
           :key   => "Name",
           :value => label
         }]
       }]
-    )
-    ec2.client.wait_until(:instance_status_ok, :instance_ids => [instance[0].id])
+    ).first
+    ec2.client.wait_until(:instance_status_ok, :instance_ids => [instance.id])
 
     _log.info("Start to load SSA application, this may take a while ...")
 
-    setup_agent(instance[0])
-    _log.info("SSA agent is ready to receive requests.")
+    setup_agent(instance)
+    _log.info("Docker #{agent_docker_name} is loaded. Start to heartbeat.")
 
-    instance[0].id
+    instance.id
   end
 
   def setup_agent(instance)
-    # register if needed
-    register(instance)
+    # Somehow instance.public_dns_name is empty, need to reinitialize to get it back
+    ip = ec2.instance(instance.id).public_dns_name || raise("Failed to get agent's public ip!")
+    key_name = instance.key_name
+    auth_key = get_keypair(key_name).try(:auth_key)
+    _log.error("Key [#{key_name}] is missing. Cannot SSH to the agent:#{instance.id}") if auth_key.nil?
+
+    ssh = LinuxAdmin::SSH.new(ip, agent_ami_login_user, auth_key)
 
     # prepare work directory
-    ssh_commands(instance, ["sudo mkdir -p #{WORK_DIR}; sudo chmod go+w #{WORK_DIR}"], agent_ami_login_user)
+    ssh.perform_commands(["sudo mkdir -p #{WORK_DIR}; sudo chmod go+w #{WORK_DIR}"])
```

Comment:
This array should be two strings, they're separate commands

Reply:
We can do it in a follow up PR